File: Ikan.cpp
#include "Ikan.h"
#include <math.h>
#include <cstdlib>

Ikan::Ikan(double x, double y, double speed, double direction, double radius) : AquariumObject(x,y,speed), HewanAir(direction, radius), hungerTimerThreshold(baseHungerThreshold), deathTimerThreshold(baseDeathThreshold) {
  hungry = false;
  timer = 0;
  timergerak = 0;
}

// GETTER & SETTER
bool Ikan::isHungry() const {
  return hungry;
}
double Ikan::getTimer() const {
  return timer;
}
void Ikan::setHungry() {
  hungry = (timer >= hungerTimerThreshold);
}
void Ikan::setTimer(double time) {
  timer = time;
}

/* METHOD SPESIFIKASI */
//Method untuk menjatuhkan koin
Koin* Ikan::coinDrop(int value) {
  Koin *coin = new Koin(this->getX(), this->getY(), value);
  return coin;
}

void Ikan::move(double timepassed) {
  double xmove = cos(direction)*speed*timepassed;
  double ymove = sin(direction)*speed*timepassed;
  this->setY(this->getY()+ymove);
  this->setX(this->getX()+xmove);
}
//bergerak secara acak
void Ikan::wander(int sizeX, int sizeY) {
  if (this->getX() + cos(direction)*speed*(5-timergerak) > sizeX || this->getX() + cos(direction)*speed*(5-timergerak) < 0 ||
		this->getY() + sin(direction)*speed*(5-timergerak) > sizeY || this->getY() + sin(direction)*speed*(5-timergerak) < 0){
	  this->setDirection(this->getDirection()+2);
  } else {
    if (timergerak > 5){
      timergerak = 0;
      this->setDirection(3.1415*cos(timer));
    }
  }
}

int Ikan::pursue(List<AquariumObject*>& food) {
  AquariumObject *p;
  int imin = 0;
  double minDist = this->distance(food.get(0));
  for(int i = 1; i < food.length(); i++){

      if (this->distance(food.get(i)) < minDist){
          imin = i;
          minDist = this->distance(food.get(i));  
      }
  }
  p = food.get(imin);
  setDirection(atan2(p->getY() - this->getY(), p->getX() - this->getX()));
  return imin;
}

bool Ikan::updateIkan(int sizeX, int sizeY, List<AquariumObject*>& food, List<AquariumObject*>& coins, double timePassed) {
  timer += timePassed;
  timergerak += timePassed;
  if (hungry && timer > deathTimerThreshold) {
    return false;
  } else if (!hungry && timer > hungerTimerThreshold) {
  	hungry = true;
  }
  return true;
  
}

File: Ikan.java
import java.util.Random;

public abstract class Ikan extends HewanAir {
    public static final int RADIUS = 50;
    private boolean hungry;
    private double timer;
    private double timergerak;
    private static final int hungerThreshold = 10;
    private static final int deathThreshold = 20;

    public Ikan(double absis, double ordinat, double speed, double direction, double radius) {
        super(absis,ordinat,speed);
        timer = 0;
        timergerak = 0;
        hungry = false;
    }
    boolean isHungry() {
        return hungry;
    }
    double getTimer(){
        return timer;
    }
    void setHungry(){
        hungry = (timer > hungerThreshold);
    }
    void setTimer(double time){
        timer = time;
    }
  
    /* METHOD SPESIFIKASI */
    //bergerak secara acak
    void wander(int sizeAbsis, int sizeOrdinat){
        if (this.getAbsis() + Math.cos(direction)*speed*(5-timergerak) > sizeAbsis || this.getAbsis() + Math.cos(direction)*speed*(5-timergerak) < 0 ||
            this.getOrdinat() + Math.sin(direction)*speed*(5-timergerak) > sizeOrdinat || this.getOrdinat() + Math.sin(direction)*speed*(5-timergerak) < 0){
                Random direct = new Random();
                this.setDirection(this.getDirection()+direct.nextFloat()*2+2);
                if (this.getDirection() > 10) {this.setDirection(this.getDirection()-2*Math.PI);}
        } else {
            if (timergerak > 5){
                timergerak = 0;
                this.setDirection(Math.PI*Math.cos(timer));
            }
        }
    }
    
    //Menjatuhkan Koin
    abstract Koin coinDrop(int value);
    
    void move(double timePassed){
        double horizontalMove = Math.cos(direction)*speed*timepassed;
        double verticalMove = Math.sin(direction)*speed*timepassed;
        this.setOrdinat(this.getOrdinat()+verticalMove);
        this.setAbsis(this.getAbsis()+horizontalMove);
    }

    int pursue(List<AquariumObject> food){
        AquariumObject pursued;
        int iMin = 0;
        double minDist = this.distance(food.get(0));
        for(int i = 1; i < food.length(); i++){
            if (this.distance(food.get(i)) < minDist){
                iMin = i;
                minDist = this.distance(food.get(i));  
            }
        }
        pursued = food.get(iMin);
        setDirection(Math.atan2(pursued.getOrdinat() - this.getOrdinat(), pursued.getAbsis() - this.getAbsis()));
        return iMin;
    }
    
    abstract int consume(List<AquariumObject> food, int indeks);

    boolean updateIkan(int sizeAbsis, int sizeOrdinat, List<AquariumObject> food, List<AquariumObject> coins, double timePassed) {
        this.timer += timePassed;
        this.timergerak += timePassed;
        if (hungry && timer > deathTimerThreshold) {
            return false;
        } else if (!hungry && timer > hungerTimerThreshold) {
            hungry = true;
        }
        return true;
    }

}